File: main-server/services/frontend/src/constants/metricTypes.ts
// SYNC: main-server/services/admin-api/services/metric_types.py
// 변경 시 양쪽 파일을 함께 수정할 것.

export const METRIC_TYPES = [
  'cpu',
  'memory',
  'disk_io',
  'network_rx',
  'network_tx',
  'http_latency',
  'log_error_rate',
] as const

export type MetricType = (typeof METRIC_TYPES)[number]

export const METRIC_TYPE_LABELS_KO: Record<MetricType, string> = {
  cpu: 'CPU 사용률',
  memory: '메모리 사용률',
  disk_io: '디스크 I/O 지연',
  network_rx: '네트워크 수신',
  network_tx: '네트워크 송신',
  http_latency: 'HTTP 응답 지연',
  log_error_rate: '로그 에러 발생률',
}

/** 메트릭 종류별 단위 — 사용자가 override_threshold 입력 시 참고용 */
export const METRIC_TYPE_UNITS: Record<MetricType, string> = {
  cpu: '%',
  memory: '%',
  disk_io: 'ms',
  network_rx: 'MB/s',
  network_tx: 'MB/s',
  http_latency: 'ms',
  log_error_rate: '건/분',
}

const TITLE_PATTERNS: { regex: RegExp; type: MetricType }[] = [
  { regex: /CPU\s*평균/, type: 'cpu' },
  { regex: /메모리\s*사용률/, type: 'memory' },
  { regex: /디스크\s*I\/?O/i, type: 'disk_io' },
  { regex: /네트워크\s*RX/i, type: 'network_rx' },
  { regex: /네트워크\s*TX/i, type: 'network_tx' },
  { regex: /HTTP\s*지연/, type: 'http_latency' },
  { regex: /로그\s*에러/, type: 'log_error_rate' },
]

/** prometheus_analyzer 알림 title에서 메트릭 종류 추출 (metric_types 컬럼 폴백). */
export function extractMetricTypesFromTitle(title: string | null | undefined): MetricType[] {
  if (!title) return []
  const found: MetricType[] = []
  for (const { regex, type } of TITLE_PATTERNS) {
    if (regex.test(title) && !found.includes(type)) found.push(type)
  }
  return found
}

/** prometheus_analyzer 알림 여부 판단 */
export function isMetricAnalyzerAlert(alert: {
  alertname: string | null
  instance_role: string | null
}): boolean {
  return (
    alert.alertname === 'prometheus_analyzer_anomaly' ||
    alert.instance_role === 'prometheus_analyzer'
  )
}

File: main-server/services/admin-api/services/metric_types.py
"""prometheus_analyzer 메트릭 종류 정의 — 예외처리(metric_exclusions) 매칭 키.

# SYNC: main-server/services/frontend/src/constants/metricTypes.ts
변경 시 양쪽 파일을 함께 수정할 것.
"""
from __future__ import annotations

import re
from enum import Enum


class MetricType(str, Enum):
    CPU = "cpu"
    MEMORY = "memory"
    DISK_IO = "disk_io"
    NETWORK_RX = "network_rx"
    NETWORK_TX = "network_tx"
    HTTP_LATENCY = "http_latency"
    LOG_ERROR_RATE = "log_error_rate"
    ZOMBIE = "zombie"


# 한국어 라벨 — UI 표시용 (DB 저장 값은 enum value 그대로 사용)
METRIC_TYPE_LABELS_KO: dict[str, str] = {
    MetricType.CPU.value: "CPU 사용률",
    MetricType.MEMORY.value: "메모리 사용률",
    MetricType.DISK_IO.value: "디스크 I/O 지연",
    MetricType.NETWORK_RX.value: "네트워크 수신",
    MetricType.NETWORK_TX.value: "네트워크 송신",
    MetricType.HTTP_LATENCY.value: "HTTP 응답 지연",
    MetricType.LOG_ERROR_RATE.value: "로그 에러 발생률",
    MetricType.ZOMBIE.value: "좀비 프로세스",
}

ALLOWED_METRIC_TYPES = frozenset(mt.value for mt in MetricType)


# anomalies title 정규식 매핑 — 레거시 alert_history.metric_types=NULL 행에서 메트릭 종류 추출용
# prometheus_analyzer.py 의 push 메시지와 매칭됨
# (예: "CPU 평균 75.0% (임계치 70%)", "디스크 I/O 278ms (임계치 200ms)")
_TITLE_PATTERNS: list[tuple[re.Pattern[str], str]] = [
    (re.compile(r"CPU\s*평균"), MetricType.CPU.value),
    (re.compile(r"메모리\s*사용률"), MetricType.MEMORY.value),
    (re.compile(r"디스크\s*I/?O"), MetricType.DISK_IO.value),
    (re.compile(r"네트워크\s*RX"), MetricType.NETWORK_RX.value),
    (re.compile(r"네트워크\s*TX"), MetricType.NETWORK_TX.value),
    (re.compile(r"HTTP\s*지연"), MetricType.HTTP_LATENCY.value),
    (re.compile(r"로그\s*에러"), MetricType.LOG_ERROR_RATE.value),
    (re.compile(r"좀비\s*프로세스"), MetricType.ZOMBIE.value),
]


def extract_metric_types_from_title(title: str) -> list[str]:
    """레거시 prometheus_analyzer 알림 title 에서 메트릭 종류 추출 (metric_types 컬럼 폴백용).

    title 예시: "[prometheus_analyzer] 디스크 I/O 278ms (임계치 200ms), CPU 평균 92.0% (임계치 70%)"
    반환: ["disk_io", "cpu"] (중복 제거, 등장 순서 유지)
    """
    if not title:
        return []
    found: list[str] = []
    for pattern, mtype in _TITLE_PATTERNS:
        if pattern.search(title) and mtype not in found:
            found.append(mtype)
    return found
